examples/07-custom-module.rs

```diff
@@ -1,49 +1,64 @@
 //! Demonstrates how to build a custom [nn::Module] without using tuples
 
 use dfdx::{
-    nn::modules::{BuildModule, Linear, Module, ReLU},
-    shapes::{Rank1, Rank2},
-    tensor::{HasErr, SampleTensor, Tape, Tensor},
+    nn::modules::{Linear, Module, ModuleVisitor, ModuleVisitorOutput, ReLU, TensorCollection},
+    prelude::BuildModule,
+    shapes::{Dtype, Rank1, Rank2},
+    tensor::{SampleTensor, Tape, Tensor},
+    tensor_ops::Device,
 };
 
 #[cfg(not(feature = "cuda"))]
-type Device = dfdx::tensor::Cpu;
+type Dev = dfdx::tensor::Cpu;
 
 #[cfg(feature = "cuda")]
-type Device = dfdx::tensor::Cuda;
-
-type Err = <Device as HasErr>::Err;
+type Dev = dfdx::tensor::Cuda;
 
 /// Custom model struct
 /// This case is trivial and should be done with a tuple of linears and relus,
 /// but it demonstrates how to build models with custom behavior
-struct Mlp<const IN: usize, const INNER: usize, const OUT: usize> {
-    l1: Linear<IN, INNER, f32, Device>,
-    l2: Linear<INNER, OUT, f32, Device>,
+struct Mlp<const IN: usize, const INNER: usize, const OUT: usize, E: Dtype, D: Device<E>> {
+    l1: Linear<IN, INNER, E, D>,
+    l2: Linear<INNER, OUT, E, D>,
     relu: ReLU,
 }
 
-// BuildModule lets you randomize a model's parameters
-impl<const IN: usize, const INNER: usize, const OUT: usize> BuildModule<Device, f32>
-    for Mlp<IN, INNER, OUT>
+// TensorCollection lets you do several operations on Modules, including constructing them with
+// randomized parameters, and iterating through or mutating all tensors in a model.
+impl<const IN: usize, const INNER: usize, const OUT: usize, E: Dtype, D: Device<E>>
+    TensorCollection<E, D> for Mlp<IN, INNER, OUT, E, D>
 {
-    fn try_build(device: &Device) -> Result<Self, Err> {
-        Ok(Self {
-            l1: BuildModule::try_build(device)?,
-            l2: BuildModule::try_build(device)?,
-            relu: ReLU,
-        })
+    // Type alias that specifies the how Mlp's type changes when using a different dtype and/or
+    // device.
+    type Output<E2: Dtype, D2: Device<E2>> = Mlp<IN, INNER, OUT, E2, D2>;
+
+    fn iter_tensors<E2: Dtype, D2: Device<E2>, V: ModuleVisitor<Self, E, D, E2, D2>>(
+        visitor: &mut V,
+    ) -> ModuleVisitorOutput<V::Func, Self, E, D, E2, D2> {
+        // Specify the name of each field and how to access it immutably and mutably.
+        //
+        // Note that 'l1' and 'l2' are of type 'Option<Linear<...>>', and that these options should
+        // not be handled until we construct the output Mlp, because a value of None is not an error.
+        let l1 = visitor.visit_module("l1", |s| &s.l1, |s| &mut s.l1)?;
+        let l2 = visitor.visit_module("l2", |s| &s.l2, |s| &mut s.l2)?;
+
+        // Specify how to construct Mlp given optional values for its fields.
+        Ok(dfdx::try_some!(Mlp {
+            l1: l1?,
+            l2: l2?,
+            relu: Default::default(),
+        }))
```

Comment:
I think the mixing of option/result may be a bit too subtle for average user. I.e. it might be too easy to mess this method up.

We should explore some alternative UX options for this. It'd be nice to have something like a builder pattern here, where users would have to always call .build() at the end:

```rust
visitor
    .module("l1", ...)
    .module("l2", ...)
    .build(|(l1, l2)| Mlp { l1, l2 relu: Default::default })
```

If we could get to something *like* that, I think we could avoid the users having to think about options/results at all, and instead handle them in the background.

I know this would be a large departure from what we have now, but it'd be a big improvement IMO

Reply:
Yeah, this is a much nicer syntax, and while your suggestion doesn't seem to be possible, it should be possible to have the following syntax:
```rust
visitor.visit_fields(
    (
        ModuleField("l1", |s| &s.l1, |s| &mut s.l1),
        ModuleField("l2", |s| &s.l2, |s| &mut s.l2),
        TensorField("example", |s| &s.example, |s| &mut s.example, TensorOptions::...),
    ),
    |(l1, l2, example)| {
        Mlp {
            l1,
            l2,
            example,
            relu: Default::default()
        }
    }
)
```
using something like the following trait:
```rust
pub trait ModuleFields<M, E: Dtype, D: Device<E>> {
    type Options<E2, D2>;
    type Output<E2, D2>;

    fn visit_fields<E2: Dtype, D2: Device, V: ModuleVisitor<M, E, D, E2, D2>>(
        &mut self,
        module: M
    ) -> Result<Self::Options<E2, D2>, Self::Err>;

    fn handle_options<E2: Dtype, D2: Device>(Self::Options<E2, D2>) -> Option<Self::Output<E2, D2>>;
}
```

Reply:
Yeah I tried for a bit yesterday to do the builder pattern, but it was very difficult 😅 

I *really* like that, nice idea! We might even be able to get rid of the visitor all together, and just have the TensorCollection with a static method that returns the fields & build closure:

```rust
   let fields = (
        ModuleField("l1", |s| &s.l1, |s| &mut s.l1),
        ModuleField("l2", |s| &s.l2, |s| &mut s.l2),
        TensorField("example", |s| &s.example, |s| &mut s.example, TensorOptions::...),
    );
    let build = |(l1, l2, example)| {
        Mlp {
            l1,
            l2,
            example,
            relu: Default::default()
        }
    };
    (fields, build)
```

Reply:
That's a really great idea, but it seems impractical, because this would either require specifying the entire type of both `fields` and `build` or trait objects, because you cannot return `impl Trait` from trait methods.

I think that the builder pattern is technically possible, but it would require something like TupleLists to function, which is less than ideal.